src/v/raft/mux_state_machine.h

```diff
@@ -214,31 +209,69 @@ ss::future<std::error_code> mux_state_machine<T...>::replicate_and_wait(
-
-                    auto last_offset = r.value().last_offset;
-
-                    auto [it, insterted] = _promises.emplace(
-                      last_offset, expiring_promise<ret_t>{});
-                    vassert(
-                      insterted,
-                      "Prosmise for offset {} already registered",
-                      last_offset);
-                    return it->second.get_future_with_timeout(
-                      timeout, [] { return errc::timeout; }, as);
-                });
-          });
+    if (_gate.is_closed()) {
+        return ss::make_ready_future<std::error_code>(errc::shutting_down);
+    }
+
+    auto u = get_units();
+    auto promise = std::make_unique<promise_t>();
+    auto f = promise->get_future_with_timeout(
+      timeout, [] { return make_error_code(errc::timeout); }, as);
+
+    ssx::spawn_with_gate(
+      _gate,
+      [this,
+       b = std::move(b),
+       u = std::move(u),
+       promise = std::move(promise)]() mutable {
+          return replicate(std::move(b))
+            .then_wrapped(
+              [this, u = std::move(u), promise = std::move(promise)](
+                ss::future<result<raft::replicate_result>> f) mutable {
+                  if (f.failed()) {
+                      promise->set_exception(f.get_exception());
+                      return ss::now();
+                  }
+                  auto r = f.get();
+                  if (!r) {
+                      promise->set_value(r.error());
+                      return ss::now();
+                  }
+
+                  auto last_offset = r.value().last_offset;
+
+                  return _new_result
+                    .wait([this, last_offset] {
+                        return _last_applied >= last_offset;
+                    })
+                    .then_wrapped(
+                      [this,
+                       last_offset,
+                       u = std::move(u),
+                       promise = std::move(promise)](ss::future<> f) mutable {
+                          u.release();
+                          if (f.failed()) {
+                              promise->set_exception(f.get_exception());
+                              return;
+                          }
+                          auto it = _results.find(last_offset);
+                          vassert(
+                            it != _results.end(),
+                            "last applied offset {} is greater than "
+                            "returned replicate result {}. this must "
+                            "imply existence of a result in results map",
+                            _last_applied,
+                            last_offset);
+                          auto res = it->second;
+                          _results.erase(it);
+                          if (_pending == 0) {
+                              _results.clear();
+                          }
+                          promise->set_value(res);
```

Comment:
I think that in a follow up patch we can simply if this by having replicate and apply play both roles: whichever fiber fires first for a given offset it adds an entry into the index and whichever comes second removes it.

Reply:
I will implement this as a follow up PR